// angular-split/esm2020/lib/module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { SplitComponent } from './component/split.component';
import { SplitAreaDirective } from './directive/split-area.directive';
import * as i0 from "@angular/core";
export class AngularSplitModule {
    static forRoot() {
        console.warn(`AngularSplitModule.forRoot() is deprecated and will be removed in v6`);
        return {
            ngModule: AngularSplitModule,
            providers: [],
        };
    }
    static forChild() {
        console.warn(`AngularSplitModule.forChild() is deprecated and will be removed in v6`);
        return {
            ngModule: AngularSplitModule,
            providers: [],
        };
    }
}
/** @nocollapse */ AngularSplitModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.0", ngImport: i0, type: AngularSplitModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
/** @nocollapse */ AngularSplitModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.0", ngImport: i0, type: AngularSplitModule, declarations: [SplitComponent, SplitAreaDirective], imports: [CommonModule], exports: [SplitComponent, SplitAreaDirective] });
/** @nocollapse */ AngularSplitModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.0", ngImport: i0, type: AngularSplitModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.0", ngImport: i0, type: AngularSplitModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [SplitComponent, SplitAreaDirective],
                    exports: [SplitComponent, SplitAreaDirective],
                }]
        }] });
